File: MainApp/src/main/java/com/sleticalboy/dailywork/weight/view/SingleLine.java
package com.sleticalboy.dailywork.weight.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.ColorRes;
import androidx.annotation.Nullable;

import com.sleticalboy.dailywork.R;

/**
 * Created on 18-3-1.
 *
 * @author leebin
 */
public class SingleLine extends View {

    private static final int DEF_MAIN_COLOR = Color.parseColor("#e5e5e5");
    private final Paint mPaint;
    private float circlePadding = 20;
    private float radius = 10;
    private float lineWidth = 4;
    private int mainColor;

    public SingleLine(Context context) {
        this(context, null);
    }

    public SingleLine(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SingleLine(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        readAttrs(context, attrs);
    }

    private void readAttrs(Context context, AttributeSet attrs) {
        if (attrs != null) {
            TypedArray a = null;
            try {
                a = context.obtainStyledAttributes(attrs, R.styleable.SingleLine);
                circlePadding = a.getDimension(R.styleable.SingleLine_circlePadding, 20);
                radius = a.getInteger(R.styleable.SingleLine_circleRadius, 10);
                lineWidth = a.getInteger(R.styleable.SingleLine_lineWidth, 4);
                mainColor = a.getColor(R.styleable.SingleLine_lineColor, DEF_MAIN_COLOR);
            } finally {
                if (a != null) {
                    a.recycle();
                }
            }

        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        mPaint.setColor(mainColor);
        mPaint.setStrokeWidth(lineWidth);
        mPaint.setStyle(Paint.Style.STROKE);
        final int cx = getWidth() >> 1;
        final int cy = getHeight() >> 1;
        canvas.drawLine(getPaddingLeft(), cy, cx - circlePadding, cy, mPaint);

        mPaint.setStyle(Paint.Style.FILL);
        canvas.drawCircle(cx, cy, radius, mPaint);

        mPaint.setStyle(Paint.Style.STROKE);
        canvas.drawLine(cx + circlePadding, cy, cx * 2 - getPaddingRight(), cy, mPaint);
    }

    public void setRadius(int radius) {
        this.radius = radius;
        invalidate();
    }

    public void setCirclePadding(int circlePadding) {
        this.circlePadding = circlePadding;
        invalidate();
    }

    public void setLineWidth(int lineWidth) {
        this.lineWidth = lineWidth;
        invalidate();
    }

    public void setMainColor(@ColorRes int mainColor) {
        this.mainColor = getResources().getColor(mainColor);
        invalidate();
    }
}

File: MainApp/src/main/java/com/sleticalboy/weight/view/SingleLine.kt
package com.sleticalboy.weight.view

import android.content.Context
import android.content.res.TypedArray
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import androidx.annotation.ColorRes
import com.sleticalboy.learning.R


/**
 * Created on 18-3-1.
 *
 * @author leebin
 */
class SingleLine @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : View(context, attrs, defStyleAttr) {
    private val mPaint: Paint
    private var circlePadding = 20f
    private var radius = 10f
    private var lineWidth = 4f
    private var mainColor = 0
    private fun readAttrs(context: Context, attrs: AttributeSet?) {
        if (attrs != null) {
            var a: TypedArray? = null
            try {
                a = context.obtainStyledAttributes(attrs, R.styleable.SingleLine)
                circlePadding = a.getDimension(R.styleable.SingleLine_circlePadding, 20f)
                radius = a.getInteger(R.styleable.SingleLine_circleRadius, 10).toFloat()
                lineWidth = a.getInteger(R.styleable.SingleLine_lineWidth, 4).toFloat()
                mainColor = a.getColor(R.styleable.SingleLine_lineColor, DEF_MAIN_COLOR)
            } finally {
                a?.recycle()
            }
        }
    }

    override fun onDraw(canvas: Canvas) {
        mPaint.color = mainColor
        mPaint.strokeWidth = lineWidth
        mPaint.style = Paint.Style.STROKE
        val cx = width shr 1
        val cy = height shr 1
        canvas.drawLine(paddingLeft.toFloat(), cy.toFloat(), cx - circlePadding, cy.toFloat(), mPaint)
        mPaint.style = Paint.Style.FILL
        canvas.drawCircle(cx.toFloat(), cy.toFloat(), radius, mPaint)
        mPaint.style = Paint.Style.STROKE
        canvas.drawLine(cx + circlePadding, cy.toFloat(), cx * 2 - paddingRight.toFloat(), cy.toFloat(), mPaint)
    }

    fun setRadius(radius: Int) {
        this.radius = radius.toFloat()
        invalidate()
    }

    fun setCirclePadding(circlePadding: Int) {
        this.circlePadding = circlePadding.toFloat()
        invalidate()
    }

    fun setLineWidth(lineWidth: Int) {
        this.lineWidth = lineWidth.toFloat()
        invalidate()
    }

    fun setMainColor(@ColorRes mainColor: Int) {
        this.mainColor = resources.getColor(mainColor)
        invalidate()
    }

    companion object {
        private val DEF_MAIN_COLOR = Color.parseColor("#e5e5e5")
    }

    init {
        mPaint = Paint()
        mPaint.isAntiAlias = true
        readAttrs(context, attrs)
    }
}